### PR TITLE
Generic ephemeral volume enablement

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

This adds further tests that ensure that the `GenericEphemeralVolume` feature works as intended. 

Handling of the new volume type was slightly broken in the kube-apiserver and didn't preserve existing instances of the type when storing a pod update.

#### Does this PR introduce a user-facing change?
```release-note
kube-apiserver: an update of a pod with a generic ephemeral volume dropped that volume if the feature had been disabled since creating the pod with such a volume
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md
```

### CODE REVIEW

**Comment by pohly** [comment]
@troy0820: this is needed for a feature which is meant to go beta in 1.21. Can we bump the priority a bit? "longterm" might be too late :sweat_smile:

File: pkg/registry/core/pod/strategy_test.go [hunk 1]
```diff
@@ -1103,3 +1103,73 @@ func TestApplySeccompVersionSkew(t *testing.T) {
...
+	updatePod := testUpdatePod(t, pod, "aaa")
+
+	// And let's enable the FG again, add another from and check if the volume is still present
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.GenericEphemeralVolume, true)()
+	testUpdatePod(t,
+		updatePod, "bbb")
+}
+
+func testUpdatePod(t *testing.T, oldPod *api.Pod, labelValue string) *api.Pod {
+	updatedPod := oldPod.DeepCopyObject().(*api.Pod)
```
Comment (gnufied) [inline pkg/registry/core/pod/strategy_test.go:1167]: Why not use `oldObject.DeepCopy` ?

File: pkg/registry/core/pod/strategy_test.go [hunk 2]
```diff
@@ -1103,3 +1103,73 @@ func TestApplySeccompVersionSkew(t *testing.T) {
...
+						},
+					},
+				},
+			},
+		},
+	}
+	expectedPod := pod.DeepCopyObject().(*api.Pod)
+
+	Strategy.PrepareForCreate(context.Background(), pod)
+	require.Equal(t, expectedPod.Spec, pod.Spec, "pod spec")
```
Comment (gnufied) [inline pkg/registry/core/pod/strategy_test.go:1151]: Do we have tests that check if feature gate is disabled then then the field is dropped?

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go [hunk 3]
```diff
@@ -454,6 +455,124 @@ func TestFindAndRemoveNonattachableVolumes(t *testing.T) {
...
+			volumeExists)
+	}
+
+	if podExistsInVolume := fakesDSW.PodExistsInVolume(
+		podName, expectedVolumeName); podExistsInVolume {
+		t.Fatalf(
+			"DSW PodExistsInVolume returned incorrect value. Expected: <false> Actual: <%v>",
+			podExistsInVolume)
+	}
+}
```
Comment (gnufied) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:574]: nit: This is bit hard to follow.  Should we have used table style tests here?

Also for last condition that we are checking, is there something special about ephemeral volumes? Seems like that is just how cleanup from DSOW behaves. Why do we have to unit test it separately?

File: pkg/registry/core/pod/strategy_test.go (hunk 1 above)
Comment (pohly) [reply pkg/registry/core/pod/strategy_test.go:1167]: Indeed, that's much simpler. Just didn't occur to me. Fixed.

File: pkg/registry/core/pod/strategy_test.go (hunk 2 above)
Comment (pohly) [reply pkg/registry/core/pod/strategy_test.go:1151]: No. I added one.

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go (hunk 3 above)
Comment (pohly) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:574]: I derived this from the other, already existing tests (for example, `TestFindAndRemoveDeletedPodsWithActualState`) and just added the feature enable/disable steps and generic ephemeral volume type instead of a persistent volume. That last check probably could be removed, but I'd prefer to keep it, just to be on the safe side and verify that using an ephemeral volume really doesn't change something compared to the persistent volume.

I don't think that a table driven approach would work better because some steps depend on the previous ones.

**Comment by pohly** [comment]
@gnufied I've pushed the update, please take another look.
